libraries/AP_RangeFinder/AP_RangeFinder_Benewake_TFMiniPlus.cpp

```diff
@@ -0,0 +1,214 @@
+
+#define DRIVER "TFMiniPlus"
+
+/*
+ * Command format:
+ *
+ * uint8_t header;
+ * uint8_t len;
+ * uint8_t id;
+ * uint8_t data[];
+ * uint8_t checksum;
+ */
+
+AP_RangeFinder_Benewake_TFMiniPlus::AP_RangeFinder_Benewake_TFMiniPlus(
+        RangeFinder::RangeFinder_State &_state,
+        AP_RangeFinder_Params &_params,
+        AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+    : AP_RangeFinder_Backend(_state, _params)
+    , _dev(std::move(dev))
+{
+}
+
+AP_RangeFinder_Backend *AP_RangeFinder_Benewake_TFMiniPlus::detect(
+        RangeFinder::RangeFinder_State &_state, AP_RangeFinder_Params &_params,
+        AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+{
+    if (!dev) {
+        return nullptr;
+    }
+
+    AP_RangeFinder_Benewake_TFMiniPlus *sensor
+        = new AP_RangeFinder_Benewake_TFMiniPlus(_state, _params, std::move(dev));
+
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+bool AP_RangeFinder_Benewake_TFMiniPlus::init()
+{
+    bool ret;
+    const uint8_t CMD_FW_VERSION[] =         { 0x5A, 0x04, 0x01, 0x5F };
+    const uint8_t CMD_SYSTEM_RESET[] =       { 0x5A, 0x04, 0x04, 0x62 };
+    const uint8_t CMD_OUTPUT_FORMAT_CM[] =   { 0x5A, 0x05, 0x05, 0x01, 0x65 };
+    const uint8_t CMD_ENABLE_DATA_OUTPUT[] = { 0x5A, 0x05, 0x07, 0x01, 0x67 };
+    const uint8_t CMD_FRAME_RATE_100HZ[] =   { 0x5A, 0x06, 0x03, 0x64, 0x00, 0xC7 };
+    const uint8_t CMD_SAVE_SETTINGS[] =      { 0x5A, 0x04, 0x11, 0x6F };
+
+    const uint8_t *cmds[] = {
+        CMD_OUTPUT_FORMAT_CM,
+        CMD_FRAME_RATE_100HZ,
+        CMD_ENABLE_DATA_OUTPUT,
+        CMD_SAVE_SETTINGS,
+    };
+    uint8_t val[12], i;
+
+    if (!_dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        hal.console->printf(DRIVER ": Unable to get bus semaphore\n");
+        return false;
+    }
+
+    _dev->set_retries(0);
+
+    /*
+     * Check we get a response for firmware version to detect if sensor is there
+     */
+    ret = _dev->transfer(CMD_FW_VERSION, sizeof(CMD_FW_VERSION), nullptr, 0);
+    if (!ret) {
+        goto fail;
+    }
+
+    hal.scheduler->delay(100);
+
+    ret = _dev->transfer(nullptr, 0, val, 7);
+    if (!ret || val[0] != 0x5A || val[1] != 0x07 || val[2] != 0x01) {
+        goto fail;
+    }
+
+    if (val[5] * 10000 + val[4] * 100 + val[3] < 10706) {
+        hal.console->printf(DRIVER ": minimum required FW version 1.7.6, but version %u.%u.%u found\n",
+                            val[5], val[4], val[3]);
+        goto fail;
+    }
+
+    hal.console->printf(DRIVER ": found fw version %u.%u.%u\n",
+                        val[5], val[4], val[3]);
+
+    for (i = 0; i < ARRAY_SIZE(cmds); i++) {
+        ret = _dev->transfer(cmds[i], cmds[i][1], nullptr, 0);
+        if (!ret) {
+            hal.console->printf(DRIVER ": Unable to set configuration register %u\n",
+                                cmds[i][2]);
+            hal.scheduler->delay(100);
+            goto fail;
```

Comment:
Those values are actually saved in the sensor. I didn't check the response as it didn't seem really important.... I'll double check later if it would make sense.